blog: How to Set Up SSO LDAP for Node-RED (#2345)
FlowFuse/website: src/blog/2024/07/how-to-setup-sso-ldap-for-the-node-red.md

LDAP enables applications and services to query, add, update, and delete directory entries stored on LDAP servers. It simplifies identity management by enabling easy authentication, authorization, and quick access to information across distributed systems.

### How LDAP SSO works

1. **User Authentication Request:** A user attempts to access a service or application that requires authentication.
   
2. **SSO Initiation:** The application forwards the authentication request to the Identity Provider (IdP) configured with LDAP, such as OpenLDAP.
   
3. **LDAP Authentication:** The IdP (LDAP server) verifies the user's credentials against its directory.
   
4. **Authentication Response:** If the credentials are valid, the LDAP server sends an authentication response (usually a token or assertion) back to the application.
   
5. **Access Granted:** The application grants access to the user based on the authentication response received from the LDAP server.

## Setting up SSO LDAP for FlowFuse

> Before we proceed, ensure that FlowFuse is deployed on your server with an Enterprise license and you have ssh connection with it so that you can run commnads in the server. If you haven't installed it yet, please check out our [documentation on installing FlowFuse](/docs/install/introduction/) or our blog on [deploying FlowFuse on Ubuntu with Docker](/blog/2024/07/deploying-flowfuse-with-docker/).

Reviewer: ```suggestion
Before we proceed, ensure that FlowFuse is deployed on your server with an Enterprise license and you have ssh connection with it so that you can run commands on the server. If you haven't installed it yet, please check out our [documentation on installing FlowFuse](/docs/install/introduction/) or our blog on [deploying FlowFuse on Ubuntu with Docker](/blog/2024/07/deploying-flowfuse-with-docker/).
```

### Installing and Configuring OpenLDAP

> Now throughout this section, we will install and configure OpenLDAP on your Ubuntu server. Make sure to replace the commands and configs with your details. If you are using a different distribution, you can follow other resources available on the internet for installation and configuration, as well as managing of users.

Reviewer: ```suggestion
Throughout this section, we will install and configure OpenLDAP on your Ubuntu server. Make sure to replace the commands and configs with your details. If you are using a different distribution, you can follow other resources available on the internet for installation and configuration, as well as managing of users.
```

1. Set the hostname for your LDAP server:

   ```bash
   hostnamectl set-hostname ldap.<your-domain>.com
   ```

2. Add the server IP to `/etc/hosts`:

   ```bash
   echo '<your_server_ip> ldap.<your-domain>.com' >> /etc/hosts
   ```

3. Install OpenLDAP and related utilities:

   ```bash
   apt install slapd ldap-utils
   ```

> 4. Set an administrator password for LDAP during installation and also conform it in the next prompt as well.

Reviewer: ```suggestion
4. Set an administrator password for LDAP during installation and confirm it in the next prompt.
```

9. Enter the Administrator password for your LDAP directory.

!["Screenshot of prompt asking to set the administrator password"](./images/Ldap-Server-Admin-Password-Ubuntu.png "Screenshot of prompt asking to set the administrator password")

10. Confirm the password.

!["Screenshot of prompt asking to conform the administrator password"](./images/Confirm-Ldap-Admin-Passsword-Ubuntu.png "Screenshot of prompt asking to conform the administrator password")

11. When asked to remove the database when slapd is purged, select ‘NO’.

!["Screenshot of the prompt asking to remove the database when slapd is purged"](./images/Database-Removal-Slapd-Ubuntu.png "Screenshot of the prompt asking to remove the database when slapd is purged")

12. Select ‘Yes’ to remove the old database to create room for a new database.

!["Screenshot of the prompt asking to remove the old database"](./images/Select-Yes-Move-Old-Databases-Slapd-Ubuntu.png "Screenshot of the prompt asking to remove the old database")

13. Edit the main OpenLDAP configuration file:

   ```bash
   sudo nano /etc/ldap/ldap.conf
   ```

> 14. Uncomment the lines beginning with “BASE” and “URI”, for example my domain is `my-flows.site`, we have updated the file as below, but you have update it according to your domain:

Reviewer: ```suggestion
14. Uncomment the lines beginning with “BASE” and “URI”, for example my domain is `my-flows.site`, we have updated the file as below, but you have to update it according to your domain:
```

```
BASE `dc=my-flows,dc=site.`
URI `ldap://ldap.my-flows.site.`
```

### Adding, Updating, and Deleting Groups and Users

#### Adding Groups and Users

1. Create a file for the base groups and open the editor:

   ```bash
   nano groups.ldif
   ```

2. Add the following content to `groups.ldif`, which will create the `users` group, make sure when you create new group the gidNumber and ou is unique:

   ```
   dn: ou=users,dc=my-flows,dc=site
   objectClass: organizationalUnit
   ou: users
   gidNumber: 7000
   ```

3. Add the groups to the LDAP directory:

   ```bash
   ldapadd -x -D cn=admin,dc=my-flows,dc=site -W -f groups.ldif
   ```

> 4. Create Password for user and store the encrypted password somewhere:

Reviewer: ```suggestion
4. Create  a password for the user and store the encrypted password:
```

!["Screenshot showing the initial form to create ldap sso configuration"](./images/sso-config-ldap.png "Screenshot showing the initial form to create ldap sso configuration")

6. In the next form, in the server field enter `your-server-ip:389`. 389 is the default port for LDAP but make sure to check it. If you are going to enable TLS, replace the port with 636.

!["Screenshot showing the advance form to create ldap sso configuration"](./images/ldap-advance-config-tag.png "Screenshot showing the advance form to create ldap sso configuration")

7. Enter the the bind DN into the username feild.
8. Enter the password for the LDAP administrator in the password feild.
9. Enter the Base DN. For example, if your domain is `my-flows.site`, the Base DN will be `dc=my-flows,dc=site`.
10. Click on the "Update configuration" button.

### Signing in Using SSO

To sign in using SSO, users of your self-hosted FlowFuse must have a FlowFuse account created with an email ID associated with the domain configured with SSO. For more information, refer to [creating users in FlowFuse](/docs/admin/user_management/#creating-new-users).

> 1. Open your platform in the browser, Enter username in the username/email feild.

Reviewer: ```suggestion
1. Open your platform in the browser. Enter the username in the username/email field.
```

!["Screenshot of prompt asking to set the administrator password while installation"](./images/Ldap-Server-Admin-Password-Ubuntu.png "Screenshot of prompt asking to set the administrator password")

!["Screenshot of prompt asking to conform the administrator password"](./images/Confirm-Ldap-Admin-Passsword-Ubuntu.png "Screenshot of prompt asking to conform the administrator password")

5. Reconfigure the slapd package:

   ```bash
   dpkg-reconfigure slapd
   ``` 

6. When asked to omit server configuration, select ‘NO’

!["Screenshot of prompt asking to omit the server configuration"](./images/Cofiguring-slapd-ubuntu.png "Screenshot of prompt asking to omit the server configuration")

7. Configure the base DN (Distinguished Name) for your LDAP directory:

   - Use your domain name to construct the base DN. For example, if your domain is `my-flows.site`, the base DN would be `dc=my-flows,dc=site` and press 'ENTER' to confirm.

![Screenshot of prompt asking to enter your domain to construct the base DN"](./images/prompt-for-domain.jpeg "Screenshot of prompt asking to enter your domain to construct the base DN")

>    - Provide a name for your organization, which will also be part of the base DN and press 'Enter'.

Reviewer: ```suggestion
   - Provide a name for your organization, which will also be part of the base DN and press 'Enter.'
```

9. Enter the Administrator password for your LDAP directory.

!["Screenshot of prompt asking to set the administrator password"](./images/Ldap-Server-Admin-Password-Ubuntu.png "Screenshot of prompt asking to set the administrator password")

10. Confirm the password.

!["Screenshot of prompt asking to conform the administrator password"](./images/Confirm-Ldap-Admin-Passsword-Ubuntu.png "Screenshot of prompt asking to conform the administrator password")

11. When asked to remove the database when slapd is purged, select ‘NO’.

!["Screenshot of the prompt asking to remove the database when slapd is purged"](./images/Database-Removal-Slapd-Ubuntu.png "Screenshot of the prompt asking to remove the database when slapd is purged")

12. Select ‘Yes’ to remove the old database to create room for a new database.

!["Screenshot of the prompt asking to remove the old database"](./images/Select-Yes-Move-Old-Databases-Slapd-Ubuntu.png "Screenshot of the prompt asking to remove the old database")

13. Edit the main OpenLDAP configuration file:

   ```bash
   sudo nano /etc/ldap/ldap.conf
   ```

> 14. Uncomment the lines beginning with “BASE” and “URI”, for example my domain is `my-flows.site`, we have updated the file as below, but you have to update it according to your domain:

Reviewer: ```suggestion
14. Uncomment the lines beginning with “BASE” and “URI”, for example, my domain is `my-flows.site`, we have updated the file as below, but you have to update it according to your domain:
```

```
BASE `dc=my-flows,dc=site.`
URI `ldap://ldap.my-flows.site.`
```

### Adding, Updating, and Deleting Groups and Users

#### Adding Groups and Users

1. Create a file for the base groups and open the editor:

   ```bash
   nano groups.ldif
   ```

> 2. Add the following content to `groups.ldif`, which will create the `users` group, make sure when you create new group the gidNumber and ou is unique:

Reviewer: ```suggestion
2. Add the following content to `groups.ldif`, which will create the `users` group, make sure when you create a new group the gidNumber and ou is unique:
```

1. Delete a user from the LDAP directory:

   ```bash
   ldapdelete -x -D cn=admin,dc=my-flows,dc=site -W "uid=sumit,ou=users,dc=my-flows,dc=site"
   ```

2. Delete a group from the LDAP directory:

   ```bash
   ldapdelete -x -D cn=admin,dc=my-flows,dc=site -W "ou=users,dc=my-flows,dc=site"
   ```
### Configuring and Enabling SSO in FlowFuse

1. To configure FlowFuse with SSO, make sure you are logged in as an administrator.
2. Go to Admin settings by clicking on the profile icon in the top-right corner and then selecting "Admin settings".

!["Screenshot showing the admin settings option in the profile icon"](./images/admin-setting-option.png "Screenshot showing the admin settings option in the profile icon")

3. Click on "Settings" from the left sidebar and switch to the SSO section.

!["Screenshot showing the sso section in the admin settings"](./images/setting's-sso-setting-section.png "Screenshot showing the sso section in the admin settings")

4. Click on the top-right "Create SSO configuration".

!["Screenshot showing the 'create sso configuration' button"](./images/create-sso-config-button.png "Screenshot showing the 'create sso configuration' button")

> 5. Enter name for your configuration, then enter domain with `@` prefix and select the "LDAP" option. Click on the "Create configuration" button.

Reviewer: ```suggestion
5. Enter the name for your configuration, then enter the domain with `@` prefix and select the "LDAP" option. Click on the "Create configuration" button.
```

!["Screenshot showing the initial form to create ldap sso configuration"](./images/sso-config-ldap.png "Screenshot showing the initial form to create ldap sso configuration")

6. In the next form, in the server field enter `your-server-ip:389`. 389 is the default port for LDAP but make sure to check it. If you are going to enable TLS, replace the port with 636.

!["Screenshot showing the advance form to create ldap sso configuration"](./images/ldap-advance-config-tag.png "Screenshot showing the advance form to create ldap sso configuration")

> 7. Enter the the bind DN into the username feild.

Reviewer: ```suggestion
7. Enter the the bind DN into the username field.
```

> 8. Enter the password for the LDAP administrator in the password feild.

Reviewer: ```suggestion
8. Enter the password for the LDAP administrator in the password field.
```

9. Enter the Base DN. For example, if your domain is `my-flows.site`, the Base DN will be `dc=my-flows,dc=site`.
10. Click on the "Update configuration" button.

### Signing in Using SSO

To sign in using SSO, users of your self-hosted FlowFuse must have a FlowFuse account created with an email ID associated with the domain configured with SSO. For more information, refer to [creating users in FlowFuse](/docs/admin/user_management/#creating-new-users).

1. Open your platform in the browser. Enter the username in the username/email field.
2. Click on "Login".

> 3. Then enter password set in the ldap directory for that user.

Reviewer: ```suggestion
3. Then enter the password set in the LDAP directory for that user.
```
